presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/ColumnMapping.java

```diff
@@ -23,10 +25,38 @@
 
 public final class ColumnMapping
 {
-    public static final PredicatePushdownController FULL_PUSHDOWN = domain -> new DomainPushdownResult(domain, Domain.all(domain.getType()));
-    public static final PredicatePushdownController PUSHDOWN_AND_KEEP = domain -> new DomainPushdownResult(domain, domain);
+    public static final int DEFAULT_DOMAIN_SIZE_THRESHOLD = 10_000;
+    public static final PredicatePushdownController FULL_PUSHDOWN = fullPushDown(DEFAULT_DOMAIN_SIZE_THRESHOLD);
+    public static final PredicatePushdownController PUSHDOWN_AND_KEEP = pushDownAndKeep(DEFAULT_DOMAIN_SIZE_THRESHOLD);
     public static final PredicatePushdownController DISABLE_PUSHDOWN = domain -> new DomainPushdownResult(Domain.all(domain.getType()), domain);
 
+    public static PredicatePushdownController pushDownAndKeep(int domainSizeThreshold)
+    {
+        return domain -> new DomainPushdownResult(domain.simplify(domainSizeThreshold), domain);
+    }
+
+    public static PredicatePushdownController fullPushDown(int domainSizeThreshold)
+    {
+        return domain -> {
+            if (getDomainSize(domain) > domainSizeThreshold) {
+                // pushdown simplified domain
+                return new DomainPushdownResult(domain.simplify(), domain);
+            }
+            else {
+                // full pushdown
+                return new DomainPushdownResult(domain, Domain.all(domain.getType()));
+            }
+        };
+    }
+
+    private static int getDomainSize(Domain domain)
```

Comment:
Should we have this in `Domain.java` instead ?

Reply:
I think size is too vague for this method to be part of `TupleDomain`

Reply:
It could be domain specific (see my comment about counting ranges twice).
however, the threshold is used both to determine _whether_ to simplify (in `fullPushdown` method) and also as _how_ to simplify via `Domain.simplify`  in `pushDownAndKeep`

If we think sizing logic is not generally applicable, we should not use single property in these two places